### ISSUE TITLE
Ill-behaved `apply` in the precence of canonical structures

### ISSUE THREAD

**Issue by robbertkrebbers** [coq/coq#6294]
#### Version

Coq 8.7.0

#### Operating system

Linux.

#### Description of the problem

In Iris we use both canonical structures and type classes, and that often leads to bad interaction where type class resolution fails in the presence of canonical structures. It seems these problems boil down to `apply` and `class_apply` failing to apply certain lemmas that should be possible to apply.

Below you can find an minimized example. I think this is the cause of quite a few problems that we are having, so it would be great if this could be fixed.

```coq
Class Op X := op : X -> X.
Class Unit X := unit : X.

Structure A := mk_A { A_car :> Type }.
Structure B := mk_B { B_car :> Type; B_op : Op B_car }.
Structure C := mk_C { C_car :> Type; C_op : Op C_car; C_u : Unit C_car }.
Existing Instance B_op.
Existing Instance C_u.

Coercion B_A (b : B) : A := mk_A (B_car b).
Coercion C_A (c : C) : A := mk_A (C_car c).
Coercion C_B (c : C) : B := mk_B (C_car c) (C_op c).

Canonical Structure B_A.
Canonical Structure C_A.
Canonical Structure C_B.

Record foo (X : Type) := Foo { foo_car : X }.
Arguments Foo {_} _.
Arguments foo_car {_} _.

Canonical Structure foo_A (a : A) : A := mk_A (foo a).

Definition foo_op (b : B) : Op (foo b) := fun x =>
  Foo (B_op _ (foo_car x)).
Canonical Structure foo_B (b : B) : B :=
  mk_B (foo b) (foo_op b).

Definition foo_u (c : C) : Unit (foo c) :=
  Foo (C_u _).
Canonical Structure foo_C (c : C) : C :=
  mk_C (foo c) (foo_op c) (foo_u c).

Lemma try1 (c : C) : Op (foo_A c).
Fail apply _. (* does not work *)
Fail apply B_op. (* does not work *)
refine (B_op _). (* works *)

(*
Cannot infer this placeholder of type "Op (foo_A c I)" (no type class instance found) in environment:
c : C
I : c -> Prop
*)
Abort.

Lemma try2 (b : C) : Op (foo_B b).
apply _. (* works *)
Qed.
```

Note that `refine` and `apply:` do work. However, that is of little help, since in the actual code I do not use `apply` myself, but type class resolution does.

**Comment by robbertkrebbers** [comment]
FWIW, a non minimized example using actual Iris:

```coq
From iris Require Import gmap.

Section foo.
  Context (A : ucmraT) `{Countable K} (i : K).

  (* When using the plain type it works *)
  Goal Op (gmap K A).
  Proof. apply _. (* Works *) Admitted.

  (* But when using the canonical strucsture, it does not *)
  Goal Op (gmapC K A).
  Proof.
  (* Note that this is just `Op (ofe_car (gmapC K A))`, i.e. it projects
  out the carrier of the canonical structure *)
  Fail apply _. (* FAILS *) Admitted.

  (* These problems appear at all kings of places, e.g.: *)
  Goal Proper (equiv ==> equiv) (singletonM i: A → gmap K A).
  Proof. apply _. Qed.

  Goal Proper (equiv ==> equiv) (singletonM i: A → gmapC K A).
  Proof. Fail apply _. Admitted.
End foo.
```

It should work (or actually, show the failure) with current Iris master and Coq 8.7.

**Comment by ejgallego** [comment]
As far as I understand the recommended workaround is to use either `apply:`, `refine` or define as done in other developments your own version of `refine _ _ _ ...`.

`apply` is in a sense deprecated as it will heavily change semantics once the Unifall branch is merged IIUC, so indeed I don't see the point of relying on it.

**Comment by robbertkrebbers** [comment]
But as I wrote, I literally cannot use `apply`. The problem is that I do not call `apply` myself, but type class search does. So, using `apply:` is not a workaround that makes sense.

**Comment by ejgallego** [comment]
Indeed, so shouldn't the bug instead be titled "Please allow the user to specify a custom tactic in type-class search"?

Maybe the bug is already around.

**Comment by Zimmi48** [comment]
Then the bug should rather be "typeclass search should not rely on the old unification engine", but @mattam82 is precisely working on this in his branch. In the meantime, it could also make sense to fix the actual reported issue if easily doable.

**Comment by robbertkrebbers** [comment]
When can we expect type class search to use the new unification engine? Is that something we can expect in 8.8? If not, it would be great if this bug could be fixed.

**Comment by RalfJung** [comment]
I tried the example in the first post with @mattam82's `unifall-apply` branch. The good news is that `apply B_op` works. The bad news is that `apply _` fails with
```
Error:
Unable to satisfy the following constraints:
In environment:
c : C

?o : "Op (foo_A c)"
```
`class_apply @B_op` fails with
```
Error:
In nested Ltac calls to "class_apply" and
"autoapply (constr) using (preident)", last call failed.
In environment
c : C
Unable to unify "Op ?M162" with "Op (foo_A c)".
```
`simple apply @B_op` fails with
```
Error: In environment
c : C
Unable to unify "Op ?b" with "Op (foo_A c)".
```
Now, I don't know what exactly the new unification is used for in that branch, and where it still uses the old unification. But it seems we are getting there :D

**Comment by mattam82** [comment]
Oh you need the unifall-Eauto (or whatever it’s named) branch then

**Comment by mattam82** [comment]
If you see a ?M it means a meta is there

**Comment by RalfJung** [comment]
Does unifall-Eauto subsume unifall-apply? Is there one branch which is the "latest" one of all the unifall branches?

**Comment by mattam82** [comment]
I think it was part of my prs but don’t know exactly which, let me check
tomorrow

**Comment by maximedenes** [comment]
I'm afraid I'll have to wait that @mattam82 is back to get a summary of the current state of unifall and see how I can help to solve this issue.

**Comment by maximedenes** [comment]
FTR, I started to work on https://github.com/coq/coq/pull/991 hoping to fix this issue.

**Comment by RalfJung** [comment]
If you need more testcases, hardly a month passes where we don't run into this -- and it is usually a pain to even realize that this is a problem (and even more so for new people just learning to work with Iris).

The latest instance: Try using `apply` or `rewrite` [here](https://gitlab.mpi-sws.org/iris/iris/blob/0a05a8c0b22b31f94e31e510c467800ea2dfc1f2/theories/base_logic/lib/proph_map.v#L137).

You can also grep the Iris master code for `#6294`, that will show 4 more places that I quickly found where we had already worked around this.

**Comment by maximedenes** [comment]
The path to a solution to this problem is #7825, then #991, and then the unifall-eauto branch of @mattam82. We are making progress, but this is a lot of work.

**Comment by RalfJung** [comment]
Thanks for the update!

**Comment by maximedenes** [comment]
Quick update: #7825 is almost ready to merge, modulo an Iris MR ;)

**Comment by RalfJung** [comment]
@maximedenes any update? That PR you mentioned last time is still open, and AFAIK it is not being blocked on Iris.

**Comment by maximedenes** [comment]
@RalfJung that PR is unfortunately blocked on the bug revealed by fiat-crypto-legacy. Let's discuss on the PR what needs to be done, hopefully Matthieu will have some ideas.

**Comment by RalfJung** [comment]
This bug also prevents us from using "non-canonical projections": when we start to reasonably configure our canonical structures to only make the type projection canonical, many proofs start to fail. Those involving `eauto` are basically impossible to fix.

So for now, we have no real choice but to disable the "redundant-canonical-projection" warning.

**Comment by maximedenes** [comment]
> This bug also prevents us from using "non-canonical projections": when we start to reasonably configure our canonical structures to only make the type projection canonical, many proofs start to fail. Those involving `eauto` are basically impossible to fix.
> 
> So for now, we have no real choice but to disable the "redundant-canonical-projection" warning.

I've been working a lot on unblocking this:
- https://github.com/coq/coq/pull/12878
- https://github.com/coq/coq/pull/12499
- https://github.com/coq/coq/pull/12872

Then, we can hopefully finish:
- https://github.com/coq/coq/pull/7825
- https://github.com/coq/coq/pull/991
- The `unifall-eauto` branch from Matthieu

And then, it will be easy. We'll get there.

**Comment by robbertkrebbers** [comment]
> And then, it will be easy. We'll get there.

That would be wonderful. Thanks @maximedenes for working on this!

**Comment by RalfJung** [comment]
@maximedenes thanks a lot. :) It is hard for a user to understand what needs to happen in the background to make this move forwards. I am sorry if my post came across as venting frustration -- while the situation is frustrating, that's nobody's fault, and everyone is doing their best to help resolve it.

I am happy to hear that steady progress is being made here and looking forward to testing the new unification once it is in a state where that is feasible. :)

**Comment by robbertkrebbers** [comment]
> I am sorry if my post came across as venting frustration -- while the situation is frustrating, that's nobody's fault, and everyone is doing their best to help resolve it.

I totally would agree with that comment!

**Comment by RalfJung** [comment]
FWIW random unification problems are currently one of the biggest roadblocks for people to get started in Iris. I run in them basically every day that I work on our Coq development. `apply in` fails, `apply` fails, `intros ?%lemma` fails -- many of these have work-arounds (such as `apply:`, or using ssreflect's rewrite to apply an implication in a hypothesis) but they are clunky and new users don't know about them.

There's currently no way to opt-in to "new" unification locally via `Set ...`, is there?
